addons/mail/static/src/messaging/utils/throttle/throttle.js

```diff
@@ -0,0 +1,426 @@
+odoo.define('mail.messaging.utils.throttle', function (require) {
+'use strict';
+
+const { makeDeferred } = require('mail.messaging.utils.deferred');
+
+/**
+ * This module define an utility function that enables throttling calls on a
+ * provided function. Such throttled calls can be canceled, flushed and/or
+ * cleared:
+ *
+ * - cancel: Canceling a throttle function call means that if a function call is
+ *   pending invocation, cancel removes this pending call invocation. It however
+ *   preserves the internal timer of the cooling down phase of this throttle
+ *   function, meaning that any following throttle function call will be pending
+ *   and has to wait for the remaining time of the cooling down phase before
+ *   being invoked.
+ *
+ * - flush: Flushing a throttle function call means that if a function call is
+ *   pending invocation, flush immediately terminates the cooling down phase and
+ *   the pending function call is immediately invoked. Flush also works without
+ *   any pending function call: it just terminates the cooling down phase, so
+ *   that a following function call is guaranteed to be immediately called.
+ *
+ * - clear: Clearing a throttle function combines canceling and flushing
+ *   together.
```

Comment:
The standard use of throttle is to just call the throttle function.
```js
const throttleFun = throttle(fun, 500);
throttleFun();
// ...
throttleFun();
```

With typing, we want to do some stuffs on throttle, like cancel a call, reset the internal timer, or both. So basically we want `.cancel()` and `.flush()` and `.clear()`.

I think you think the API looks complicated because the comments try to explain this concept in depth, but I genuinely think this is the easiest API you could get.

Reply:
Every time I will see those method, I'll have to double check what they do if I don't remember.
Instead if I see `pending: "execute"`, `pending: "cancel"` and `timer: "reset"` I'll immediately know what it does.

Reply:
Could you provide an example of such API? I don't see at all how this would look like.

Reply:
I don't care too much about this one, but as an example I guess it would look more verbose but just do the same thing at the end:

```js
throttleFun.update({
    pending: "cancel",
    timer: "reset",
});
```
or
```js
throttleFun.update({
    cancelPending: True,
    resetTimer: True
});
```

Reply:
try with throttle and timer to see if you find something better with what we discussed